Tripgenerator

chrisvdwerf opened the pull request:
**Allows using sumo/tools/randomTrips .py to generate trips**. Ignores following parameters if `'trips_generate` is set to True:
```
'car_pr': 0.5,  # for automatic route/config generation probability that a car appears
'car_tm': 2,  #  for automatic route/config generation when the first car appears?
'route_starts' : [],  #  for automatic route/config generation, ask Rolf
'route_min_segments' : 0,  #  for automatic route/config generation, ask Rolf
'route_max_segments' : 0,  #  for automatic route/config generation, ask Rolf
'route_ends' : [],  #  for automatic route/config generation, ask Rolf
'seed': None
```

**Fixes an issue with 'g' in traffic phases**
For most custom scenario imported using osmWebWizard there are also some phases for traffic lights that are 'g' instead of 'G'. This has something to do with lane priority. The wrapper sometimes crashes when it finds a 'g' in a phase. This is fixed in this branch.

chrisvdwerf committed:
Allow using sumo/tools/randomTrips.py

chrisvdwerf committed:
Revert changes to file deletion

chrisvdwerf committed:
reorder paramameters

chrisvdwerf committed:
reorder parameters

chrisvdwerf committed:
Fix issue with 'g' in phases

chrisvdwerf committed:
Allow specifying own scenario directory

chrisvdwerf committed:
Fix path issue

chrisvdwerf committed:
change reward

chrisvdwerf committed:
Add seed possibility for randomTrips

chrisvdwerf committed:
Revert "change reward"

This reverts commit a816fc264e308be64341cad7f9e9d6f7167c94c0.

Commit:
Remove unused imports

Commit:
Add get_net_file method to SumoGymAdapter

Commit:
Note that net file can be used for tlphases

EmanuelKuhn committed:
Use netfiles for tlphases by default in SumoGymAdapter

Update GridSumoEnv to show example of using netfiles for tl phases.

EmanuelKuhn committed:
Add option to initialize State later in SumoGymAdapter

This is needed for a subclass to change parameters used to initialize
the state after __init__ is called.

EmanuelKuhn committed:
Add NetFileUtil to parse junctions in a .net.xml file

Used to map a junction name to a traffic light (tlLogic) id.

EmanuelKuhn committed:
Add class for custom OSM scenario

EmanuelKuhn committed:
Replace 'trips_generate' flag with 'route_generation_method' option

EmanuelKuhn committed:
Add stats based activitygen route generation method

EmanuelKuhn committed:
Remove print statement

EmanuelKuhn commented:
**Add CustomSumoEnv class, which allows for an imported (OSM) scenario**
  - Given a junction_id (e.g. "Waring_Woodhull"), a box around the intersection with `box_width` and `box_height` is used for the observation and reward
  - The action_space property only returns the actions of the traffic light of junction_id

  - The box is not perfectly centred around the intersection, this might be because of the coordinates in the net file

**Add activitygen as a route generation option**
  - Set the `route_generation_method` flag to 'activitygen', see below
  - This method has the drawback that the generated routes take a whole day (86400 steps), thus one episode takes a long time to simulate
  - The current statistics/parameters used generate a very small number of cars

**Replace flag `trips_generate` flag with `'route_generation_method'`**
As we also wanted to try stat based activity generation, I replaced the 'trips_generate' flag with a `'route_generation_method'`.

Valid options to choose are: ['legacy', 'randomTrips.py', 'activitygen']

EmanuelKuhn committed:
Add rp-scenario-v2 to scenarios folder

chrisvdwerf committed:
set sensible default option

chrisvdwerf commented:
Maybe an idea to allow adding command line arguments for route generation using the stats method?

chrisvdwerf commented on aienvs/Sumo/SumoCustomEnv.py:
```diff
@@ -0,0 +1,73 @@
+    __DEFAULT_PARAMETERS = {
+        'scenarios_path': os.path.join(os.path.dirname(__file__), "../../scenarios/Sumo/"),
+        'scene': 'rp-scenario-v2',
+        'trips_generate': True,  # using sumo/tools/randomTrips.py to generate trips
```
This line can be removed now right?

chrisvdwerf commented:
![image](https://user-images.githubusercontent.com/18056780/81720681-068eca00-947f-11ea-95d3-5cf90c2041d6.png)

The scene path seems to be added twice at the moment

EmanuelKuhn committed:
Add activitygen options

EmanuelKuhn committed:
Merge remote-tracking branch 'origin/tripgenerator' into tripgenerator

EmanuelKuhn committed:
Remove unused parameter

EmanuelKuhn committed:
Fix tlphasesfile path

EmanuelKuhn commented:
> ![image](https://user-images.githubusercontent.com/18056780/81720681-068eca00-947f-11ea-95d3-5cf90c2041d6.png)
> 
> The scene path seems to be added twice at the moment

I didn't notice this bug because it worked on my os, but it should be fixed now.

YanivO1123 committed:
Fix to GridSumoEnv.py to be able to deal with paths with spaces, fix to import in aienvs/SimulatedEnvironment.py

EmanuelKuhn commented on aienvs/Sumo/SumoGymAdapter.py:
```diff
@@ -27,30 +25,43 @@ class SumoGymAdapter(Env):
+                'activitygen_options': [], # e.g. ["--end", endtime]
+
+                # Options for 'route_generation_method' 'randomTrips.py'
+                'trips_generate_options': ["--validate"], # sumo/tools/randomTrips.py additional options. -n, -o already handled!
```
Is it a good idea to add options here? If you want to add extra options, "--validate" would get removed.

EmanuelKuhn commented on aienvs/Sumo/SumoHelper.py:
```diff
@@ -125,6 +122,65 @@ def _generate_route_file(self, seed):
+
+            subprocess.call(activitygen_args)
+
+            duarouter_args = [DUAROUTER, '--net-file', net_file, '--route-files', tmp_trips_route_file,
```
The random trips method only generates trips instead of also routing the cars. Only calling activitygen and not also duarouter would give the same behaviour.

If only trips are supplied to sumo, [routes are calculated when a vehicle departs](https://sumo.dlr.de/docs/Demand/Dynamic_User_Assignment.html#oneshot-assignment) taking traffic jams into account, so it's probably a good idea to also only supply trips for the activitygen method.

EmanuelKuhn replied:
I changed it to using trips directly.

chrisvdwerf committed:
Default usage of --validate

EmanuelKuhn committed:
Fix --seed of activitygen

Also make --seed more explicit for randomtrips

EmanuelKuhn committed:
Use --random for routes if seed is None

EmanuelKuhn committed:
Fix typo: trips were overwritten by routes

EmanuelKuhn committed:
Use activitygen directly with Sumo instead of via duarouter

If only trips are supplied to sumo instead of routed vehicles,
Sumo uses oneShot-assignment to route vehicles on departure.

See: https://sumo.dlr.de/docs/Demand/Dynamic_User_Assignment.html#oneshot-assignment

This should have better performance than using duarouter, as traffic jams
are taken into account. This is also what is currently used for the
randomTrips.py method.
